src/server/routers/_app.ts

```diff
@@ -16,8 +16,139 @@ import {
   publicProcedure,
   router,
 } from 'src/server/trpc'
+import { ProposalStatusFilter } from 'src/types/app'
 import { z } from 'zod'
 
+async function getStudentProposals({ ctx, filters }) {
+  const proposals = await prisma.proposal.findMany({
+    where: {
+      typeKey: ProposalType.SUPERVISOR,
+      statusKey: ProposalStatus.OPEN,
+    },
+    include: {
+      attachments: true,
+      topicArea: true,
+      supervisedBy: {
+        include: {
+          supervisor: true,
+        },
+      },
+    },
+  })
+
+  return proposals.map((p) => ({
+    ...p,
+    supervisedBy: p.supervisedBy?.[0]?.supervisor,
+  }))
+}
+
+async function getSupervisorProposals({ ctx, filters }) {
+  let where = {}
+  let applications = undefined
+  let receivedFeedbacks = undefined
+
+  if (ctx.user?.role === UserRole.SUPERVISOR) {
+    where = {
+      ...where,
+      typeKey: {
+        in: ['SUPERVISOR', 'STUDENT'],
+      },
+    }
+    applications = {
+      include: {
+        attachments: true,
+        status: true,
+      },
+    }
+    receivedFeedbacks = {
+      where: {
+        user: {
+          email: ctx.user.email,
+        },
+      },
+    }
+  } else {
+    where = {
+      ...where,
+      typeKey: {
+        in: ['SUPERVISOR'],
+      },
+    }
+  }
+
+  if (filters.status === ProposalStatusFilter.ALL_PROPOSALS) {
+    where = {
+      ...where,
+      OR: [
+        { statusKey: ProposalStatus.OPEN },
+        {
+          supervisedBy: {
+            some: {
+              supervisorEmail: ctx.user?.email,
+            },
+          },
+        },
+      ],
+    }
+  }
+
+  if (filters.status === ProposalStatusFilter.OPEN_PROPOSALS) {
+    where = {
+      ...where,
+      statusKey: ProposalStatus.OPEN,
+      NOT: {
+        receivedFeedbacks: {
+          some: {
+            userEmail: ctx.user?.email,
+          },
+        },
+      },
+    }
+  }
+
+  if (filters.status === ProposalStatusFilter.MY_PROPOSALS) {
+    where = {
+      ...where,
+      supervisedBy: {
+        some: {
+          supervisorEmail: ctx.user?.email,
+        },
+      },
+    }
+  }
+
+  if (filters.status === ProposalStatusFilter.REJECTED_n_DECLINED_PROPOSALS) {
+    where = {
+      ...where,
+      receivedFeedbacks: {
+        some: {
+          userEmail: ctx.user?.email,
+          typeKey: {
+            startsWith: 'REJECTED' || 'DECLINED',
```

Comment:
does this work? I think it would need to be an array, as || is a JS operator and probably evaluated early. would be even better to use the "in" operator with an array for exact matching (in: ["REJECTED...", "DECLINED..."])

Reply:
It works, but IDK if it's the best solution? Should I change it?

Reply:
done